### PR TITLE
[MM-20684] Slash Command Autocomplete

### PR DESCRIPTION
**Pull request by iomodo**
#### Summary
This PR implements slash command autocomplete suggestions for plugins

#### Ticket Link 
Fixes https://mattermost.atlassian.net/browse/MM-18098
Fixes https://mattermost.atlassian.net/browse/MM-21485
Fixes https://mattermost.atlassian.net/browse/MM-21486
Fixes https://mattermost.atlassian.net/browse/MM-21487
Fixes https://mattermost.atlassian.net/browse/MM-21488
Fixes https://mattermost.atlassian.net/browse/MM-21489
Fixes https://mattermost.atlassian.net/browse/MM-21490
Fixes https://mattermost.atlassian.net/browse/MM-21491
Fixes https://mattermost.atlassian.net/browse/MM-21493
Fixes https://mattermost.atlassian.net/browse/MM-21494
Fixes https://mattermost.atlassian.net/browse/MM-24766
Fixes https://mattermost.atlassian.net/browse/MM-24767

#### Related PRs
webapp - https://github.com/mattermost/mattermost-webapp/pull/5499
redux - https://github.com/mattermost/mattermost-redux/pull/1121
demo-plugin - https://github.com/mattermost/mattermost-plugin-demo/pull/99
api-reference - https://github.com/mattermost/mattermost-api-reference/pull/522

### CODE REVIEW

File: model/command_autocomplete.go [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+		HelpText: helpText,
+		Type:     AutocompleteArgTypeText,
+		Required: required,
+		Data:     &AutocompleteTextArg{Hint: hint, Pattern: pattern},
+	}
+	ad.Arguments = append(ad.Arguments, &argument)
+}
+
+// AddStaticListArgument adds positional AutocompleteArgTypeStaticList argument to the command.
+func (ad *AutocompleteData) AddStaticListArgument(helpText string, items []AutocompleteListItem, required bool) {
```
Comment (hanzei) [inline model/command_autocomplete.go:138]: Subjective nit pick: I seams to be that slice arguments usually are the last ones past to a function. Hence, I would go with:
```suggestion
func (ad *AutocompleteData) AddStaticListArgument(helpText string, required bool, items []AutocompleteListItem) {
```

File: model/command_autocomplete.go [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+	ad.AddNamedStaticListArgument("", helpText, items, required)
+}
+
+// AddNamedStaticListArgument adds named AutocompleteArgTypeStaticList argument to the command.
+func (ad *AutocompleteData) AddNamedStaticListArgument(name, helpText string, items []AutocompleteListItem, required bool) {
```
Comment (hanzei) [inline model/command_autocomplete.go:143]: Same here

File: plugin/client_rpc.go [hunk 3]
```diff
@@ -99,6 +99,9 @@ func init() {
 	gob.Register(&model.AppError{})
 	gob.Register(&ErrorString{})
 	gob.Register(&opengraph.OpenGraph{})
+	gob.Register(&model.AutocompleteDynamicListArg{})
+	gob.Register(&model.AutocompleteStaticListArg{})
+	gob.Register(&model.AutocompleteTextArg{})
```
Comment (hanzei) [inline plugin/client_rpc.go:104]: Should `AutocompleteData` also be added to the list?

File: model/command.go [hunk 4]
```diff
@@ -18,23 +18,25 @@ const (
...
+	Username             string            `json:"username"`
+	IconURL              string            `json:"icon_url"`
+	AutoComplete         bool              `json:"auto_complete"`
+	AutoCompleteDesc     string            `json:"auto_complete_desc"`
+	AutoCompleteHint     string            `json:"auto_complete_hint"`
+	DisplayName          string            `json:"display_name"`
+	Description          string            `json:"description"`
+	URL                  string            `json:"url"`
+	AutocompleteData     *AutocompleteData `db:"-" json:"autocomplete_data,omitempty"`
+	AutocompleteIconData string            `db:"-" json:"autocomplete_icon_data,omitempty"`
```
Comment (hanzei) [inline model/command.go:39]: What is the data type here and can a comment make that clear?

File: app/command_channel_rename.go [hunk 5]
```diff
@@ -25,12 +25,15 @@ func (me *RenameProvider) GetTrigger() string {
 }
 
 func (me *RenameProvider) GetCommand(a *App, T goi18n.TranslateFunc) *model.Command {
+	renameAutocompleteData := model.NewAutocompleteData(CMD_RENAME, T("api.command_channel_rename.hint"), T("api.command_channel_rename.desc"))
+	renameAutocompleteData.AddTextArgument(T("api.command_channel_rename.hint"), "[text]", "")
```
Comment (hanzei) [inline app/command_channel_rename.go:29]: There are also other commands like `code`that need to be updated. Is there a ticket to do that?

File: app/command_autocomplete.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	// input with a single word
+	index := strings.Index(in, " ")
+	if index == -1 { // typing of the single word argument is not finished
+		return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: parsed + toBeParsed, Suggestion: "", Hint: a.Hint, Description: arg.HelpText}
+	}
+	// single word argument already typed
+	return false, parsed + in[:index+1], in[index+1:], model.AutocompleteSuggestion{}
+}
+
+func parseStaticListArgument(arg *model.AutocompleteArg, parsed, toBeParsed string) (found bool, alreadyParsed string, yetToBeParsed string, suggestions []model.AutocompleteSuggestion) {
```
Comment (hanzei) [inline app/command_autocomplete.go:215]: Are the names return values needed? Maybe it make the code easier to read if they get dropped, given that the method signature is already quite long. The same comment applies to all other methods in `app/command_autocomplete.go`.

File: app/command_autocomplete.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+	a := arg.Data.(*model.AutocompleteStaticListArg)
+	return parseListItems(a.PossibleArguments, parsed, toBeParsed)
+}
+
+func (a *App) parseDynamicListArgument(arg *model.AutocompleteArg, parsed, toBeParsed string) (found bool, alreadyParsed string, yetToBeParsed string, suggestions []model.AutocompleteSuggestion) {
+	dynamicArg := arg.Data.(*model.AutocompleteDynamicListArg)
+	params := url.Values{}
+	params.Add("user_input", parsed+toBeParsed)
+	params.Add("parsed", parsed)
+	resp, err := a.doPluginRequest("GET", dynamicArg.FetchURL, params, nil)
```
Comment (hanzei) [inline app/command_autocomplete.go:225]: Maybe it's just the function name that is confusing me, but normal integrations will be able to register for a dynamic list, right?

File: app/command_autocomplete.go [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+	// single word argument already typed
+	return false, parsed + in[:index+1], in[index+1:], model.AutocompleteSuggestion{}
+}
+
+func parseStaticListArgument(arg *model.AutocompleteArg, parsed, toBeParsed string) (found bool, alreadyParsed string, yetToBeParsed string, suggestions []model.AutocompleteSuggestion) {
+	a := arg.Data.(*model.AutocompleteStaticListArg)
+	return parseListItems(a.PossibleArguments, parsed, toBeParsed)
+}
+
+func (a *App) parseDynamicListArgument(arg *model.AutocompleteArg, parsed, toBeParsed string) (found bool, alreadyParsed string, yetToBeParsed string, suggestions []model.AutocompleteSuggestion) {
```
Comment (hanzei) [inline app/command_autocomplete.go:220]: 0/5 I doesn't seam like there is much parsing
```suggestion
func (a *App) getDynamicListArgument(arg *model.AutocompleteArg, parsed, toBeParsed string) (found bool, alreadyParsed string, yetToBeParsed string, suggestions []model.AutocompleteSuggestion) {
```

File: model/command_autocomplete.go [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+
+	"github.com/pkg/errors"
+)
+
+// Argument types
+const (
+	AutocompleteArgTypeText        = "TextInput"
+	AutocompleteArgTypeStaticList  = "StaticList"
+	AutocompleteArgTypeDynamicList = "DynamicList"
+)
```
[the remark this answers is not in the document]
Comment (iomodo) [reply model/command_autocomplete.go:22]: Yes it is idiomatic, but I don't see any advantages in introducing type which is never used.

File: model/command_autocomplete.go [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+	// Complete describes completed suggestion
+	Complete string
+	// Suggestion describes what user might want to input next
+	Suggestion string
+	// Hint describes a hint about the suggested input
+	Hint string
+	// Description of the command or a suggestion
+	Description string
+	// icon
+	IconData string
```
[the remark this answers is not in the document]
Comment (iomodo) [reply model/command_autocomplete.go:100]: yes base64 encoded svgs

File: model/command_autocomplete.go [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+		Hint:        hint,
+		HelpText:    helpText,
+		RoleID:      SYSTEM_USER_ROLE_ID,
+		Arguments:   []*AutocompleteArg{},
+		SubCommands: []*AutocompleteData{},
+	}
+}
+
+// AddCommand adds a subcommand to the autocomplete data.
+func (ad *AutocompleteData) AddCommand(command *AutocompleteData) {
```
[the remark this answers is not in the document]
Comment (iomodo) [reply model/command_autocomplete.go:116]: `AddCommand` is idiomatic here, as is `addTree`(not addSubTree). For example you can see `cobra` [implementation](https://github.com/spf13/cobra/blob/94a87a7b83bda98e918cf7a4793f41372a5e4139/command.go#L1102)

File: model/command_autocomplete.go [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+	IconData string
+}
+
+// NewAutocompleteData returns new Autocomplete data.
+func NewAutocompleteData(trigger, hint, helpText string) *AutocompleteData {
+	return &AutocompleteData{
+		Trigger:     trigger,
+		Hint:        hint,
+		HelpText:    helpText,
+		RoleID:      SYSTEM_USER_ROLE_ID,
```
[the remark this answers is not in the document]
Comment (iomodo) [reply model/command_autocomplete.go:109]: There is not through the API, I think we should add that after we finalize roles and permissions.

File: model/command_autocomplete.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (iomodo) [reply model/command_autocomplete.go:116]: I tried that, but API was not good. Added `nil` checks are in the `IsValid` method instead.

File: plugin/client_rpc.go (hunk 3 above)
Comment (iomodo) [reply plugin/client_rpc.go:104]: No, this is needed only for objects which are provided using `interface{}` type

File: app/command_channel_rename.go (hunk 5 above)
Comment (iomodo) [reply app/command_channel_rename.go:29]: Yes https://mattermost.atlassian.net/browse/MM-21499. We should make them HW.

File: app/command_autocomplete.go (hunk 6 above)
Comment (iomodo) [reply app/command_autocomplete.go:215]: They are needed to understand what those returned strings will be, for readability.

File: app/command_autocomplete.go (hunk 7 above)
Comment (iomodo) [reply app/command_autocomplete.go:225]: They will be in the future, it is not implemented yet.

File: model/command_autocomplete.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply model/command_autocomplete.go:103]: Could you please make this more clear in the comment?

File: api4/command.go [hunk 13]
```diff
@@ -369,6 +370,39 @@ func listAutocompleteCommands(c *Context, w http.ResponseWriter, r *http.Request
...
+
+	commands, err := c.App.ListAutocompleteCommands(c.Params.TeamId, c.App.T)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	suggestions := c.App.GetSuggestions(commands, userInput, roleId)
+
+	w.Write(model.AutocompleteSuggestionsToJSON(suggestions))
```
Comment (levb) [inline api4/command.go:403]: set the content-type?

File: app/integration_action.go [hunk 14]
```diff
@@ -325,27 +325,34 @@ func (w *LocalResponseWriter) WriteHeader(statusCode int) {
 	w.status = statusCode
 }
 
-func (a *App) DoLocalRequest(rawURL string, body []byte) (*http.Response, *model.AppError) {
+func (a *App) doPluginRequest(method, rawURL string, values url.Values, body []byte) (*http.Response, *model.AppError) {
```
Comment (levb) [inline app/integration_action.go:328]: I am a little confused - is this a relevant change, or a rebase/merge issue? This file doesn't seem relevant to autocomplete.

File: model/command_autocomplete.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (levb) [reply model/command_autocomplete.go:103]: perhaps just `s/IconData/IconSVG[Base64]/`?

File: api4/command.go (hunk 13 above)
Comment (iomodo) [reply api4/command.go:403]: `json` is the default content-type as implemented in the handler. So no need to add it here.

File: app/integration_action.go (hunk 14 above)
Comment (iomodo) [reply app/integration_action.go:328]: It is relevant for downloading dynamic data from plugins. I just reorganized code so we could have different http methods `GET` , `POST`..

File: model/command_autocomplete.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (iomodo) [reply model/command_autocomplete.go:103]: We already use `IconData` in other parts of the code so I'll stick to the name and add a comment.

File: model/command_autocomplete.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply model/command_autocomplete.go:119]: I'm fine with the current implementation.

[the remark this answers is not in the document]
Comment (hanzei) [reply model/command_autocomplete.go:119]: Could you elaborate what issue the API without a pointer is?

File: model/command_autocomplete.go [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+	// Hint of a command
+	Hint string
+	// Text displayed to the user to help with the autocomplete description
+	HelpText string
+	// Role of the user who should be able to see the autocomplete info of this command
+	RoleID string
+	// Arguments of the command. Arguments can be named or positional.
+	// If they are positional order in the list matters, if they are named order does not matter.
+	// All arguments should be either named or positional, no mixing allowed.
+	Arguments []*AutocompleteArg
```
[the remark this answers is not in the document]
Comment (hanzei) [reply model/command_autocomplete.go:40]: I'm still think this a valid point, but I'm also fine with addressing it in a separate ticket.

File: app/command_autocomplete.go (hunk 6 above)
Comment (hanzei) [reply app/command_autocomplete.go:215]: Well, I think comes down to personal preference. I'm fine with keeping the named returns.

**Comment by DHaussermann** [comment]
Regression tested this for backwards compatibility 
- Ran through release testing of all system slash commands 
- Tested main input and RHS 
-  Regression tested other auto-completes such such as channel and @ mentions as a precaution
- Regression tested a couple plugins for backwards compatibility. No issues found passing parameters (string and bool)
-  Regression tested a custom slash command for backwards compatibility
- Confirmed with Shota web traffic on auto-complete looks as expected
- Confirmed with Shota that parsing the commands should not cause any noticeable performance deterioration 
LGTM! Please merge.

**Comment by mickmister** [comment]
@iomodo Looks like tests are failing here

**Comment by aaronrothschild** [comment]
@iomodo Let's discuss Documentation during our next 1:1 so it's ready for release.
